packages/firestore/src/local/indexeddb_mutation_queue.ts

```diff
@@ -89,45 +89,26 @@ export class IndexedDbMutationQueue implements MutationQueue {
   }
 
   start(transaction: PersistenceTransaction): PersistencePromise<void> {
-    let nextBatchId;
-
+    let globalNextBatchId;
     return IndexedDbMutationQueue.loadNextBatchIdFromDb(transaction)
-      .next(batchId => {
-        nextBatchId = batchId;
+      .next(maxBatchId => {
+        globalNextBatchId = maxBatchId;
         return mutationQueuesStore(transaction).get(this.userId);
       })
       .next((metadata: DbMutationQueue | null) => {
         if (!metadata) {
           metadata = new DbMutationQueue(
             this.userId,
-            BATCHID_UNKNOWN,
+            /*lastAcknowledgedBatchId=*/ BATCHID_UNKNOWN,
             /*lastStreamToken=*/ '',
-            BATCHID_UNKNOWN
+            /*nextBatchId=*/ BATCHID_UNKNOWN
           );
         }
-
-        // TODO(multitab): This should only run when no other client is
-        // active, as batch IDs may otherwise go back in time in clients that
-        // have already processed mutations.
-
-        // On restart, nextBatchId may end up lower than
-        // lastAcknowledgedBatchId since it's computed from the queue
-        // contents, and there may be no mutations in the queue. In this
-        // case, we need to reset lastAcknowledgedBatchId (which is safe
-        // since the queue must be empty).
-        if (metadata.lastAcknowledgedBatchId >= nextBatchId) {
-          return this.checkEmpty(transaction).next(empty => {
-            assert(
-              empty,
-              'Reset nextBatchID is only possible when the queue is empty'
-            );
-
-            metadata.lastAcknowledgedBatchId = BATCHID_UNKNOWN;
-            metadata.lastProcessedBatchId = BATCHID_UNKNOWN;
-            return mutationQueuesStore(transaction).put(metadata);
-          });
-        } else {
+        if (globalNextBatchId > metadata.nextBatchId) {
```

Comment:
Hrm...  So I think this code still allows for re-use of batchIDs, since you might have:

1. user1 writes batchid 0 (user1 metadata.nextBatchId=1)
2. user1 acks batchid 0 (so nothing in the mutations store)
3. user1 logs out and new user2 logs in.
4. user2 metadata.nextBatchId will be initialized to 0 since there's no metadata for user2 and no mutations in the queue.
5. user2 writes batchId 0 (duplicate)

This is probably non-harmful since it's okay to "reset" the batchId if there are no mutations...  but I think your intention was to never reuse batch ids, right?  And I don't think this code does that.

To fix this you need to either iterate over all users' metadata and find the max() of their nextBatchIds... or introduce a global nextBatchId in IndexedDb (as I mention elsewhere).